### PR TITLE
:sparkles: Adding OpenTelemetry Tracing for easier debugging of the moving pieces

### PR DESCRIPTION
**Pull request by shawn-hurley**
This PR just sets up the basic infrastructure but does not fully address all the things that we want. 

This is able to see the call stack for run_demo, but the SPAN is not piped through the RPC call. 

This is able to see the kai-rpc server get the request, determine what to do with it, and call the code_plan action. 

I have not instrumented any of the agents, task_runners, or model providers. I think this needs to be followed up on.

### CODE REVIEW

File: example/run_demo.py [hunk 1]
```diff
@@ -111,8 +121,12 @@ def initialize_rpc_server() -> Generator[JsonRpcServer, None, None]:
             )
         yield rpc_server
     finally:
-        rpc_subprocess.terminate()
+        # send shutdown
+        response = rpc_server.send_request("shutdown", params={})
+        log.info(f"shutdown resposne -- {response}")
```
Comment (shawn-hurley) [inline example/run_demo.py:126]: make this debug

File: kai/jsonrpc/core.py [hunk 2]
```diff
@@ -217,51 +220,68 @@ def __init__(
         self.outstanding_requests: set[JsonRpcId] = set()
 
         self.shutdown_flag = False
+        self.log = get_logger("jsonrpc-server")
+        if log is not None:
+            self.log = log
 
     def stop(self) -> None:
+        self.log.info("JsonRpcServer shutdown flag")
```
Comment (shawn-hurley) [inline kai/jsonrpc/core.py:228]: remove or make debug

File: kai/jsonrpc/core.py [hunk 3]
```diff
@@ -217,51 +220,68 @@ def __init__(
...
+                        cond = self.event_dict[msg.id]
+                        cond.acquire()
+                        cond.notify()
+                        cond.release()
+
+                else:
+                    span.add_event("received_unknown", attributes={"message": f"{msg}"})
+                    self.log.error(f"Unknown message type: {type(msg)}")
+
+        self.log.info("here ending run")
```
Comment (shawn-hurley) [inline kai/jsonrpc/core.py:284]: remove

File: kai/jsonrpc/streams.py [hunk 4]
```diff
@@ -45,8 +47,12 @@ def __init__(
         self.json_loads_kwargs = json_loads_kwargs
 
     def close(self) -> None:
-        self.recv_file.close()
-        self.send_file.close()
+        pass
```
Comment (shawn-hurley) [inline kai/jsonrpc/streams.py:50]: figure out what this should do

File: kai/jsonrpc/streams.py [hunk 5]
```diff
@@ -160,19 +166,24 @@ def __init__(
...
     ):
         super().__init__(recv_file, send_file, json_dumps_kwargs, json_loads_kwargs)
 
         self.buffer: str = ""
         self.decoder = json.JSONDecoder()
         self.chunk_size = 512
+        self.log = get_logger("jsonrpc")
+        if log is not None:
+            self.log = log
+        self.readline_task: Future[bytes | None] | None = None
```
Comment (shawn-hurley) [inline kai/jsonrpc/streams.py:179]: remove we don't need this here now

**Review by pranavgaikwad — APPROVED** [review]

**Pull request merged**